### PR TITLE
add ref_extern_data_dims_via_global_config flag

### PR DESCRIPTION
**Pull request by JackTemaki**
I changed `self._base_extern_data_dim_refs` from `set` to `list` because otherwise I would get undefined behavior. Maybe I overlooked something but this is how it worked for me now.

### CODE REVIEW

**Comment by albertz** [comment]
> I changed `self._base_extern_data_dim_refs` from `set` to `list` because otherwise I would get undefined behavior. Maybe I overlooked something but this is how it worked for me now.

I don't understand. Why? This is wrong. It should be a `set`. If there is some other bug somewhere, and this is a workaround (which I don't understand), then: We should not add workarounds but fix the real bugs instead. Or have a very good explanation why this is needed. Not just "it worked for me now".

**Comment by JackTemaki** [comment]
> Or have a very good explanation why this is needed.

A set has no defined order. Sisyphus needs a defined order as otherwise the serialized string always changes, and with it the hashes. Thus I changed this to list. Otherwise we need to add some explicit hash handling.

**Comment by albertz** [comment]
> > Or have a very good explanation why this is needed.
> 
> A set has no defined order. Sisyphus needs a defined order as otherwise the serialized string always changes, and with it the hashes. Thus I changed this to list. Otherwise we need to add some explicit hash handling.

I don't understand. `_base_extern_data_dim_refs` is never serialized.

**Comment by JackTemaki** [comment]
> > > Or have a very good explanation why this is needed.
> > 
> > 
> > A set has no defined order. Sisyphus needs a defined order as otherwise the serialized string always changes, and with it the hashes. Thus I changed this to list. Otherwise we need to add some explicit hash handling.
> 
> I don't understand. `_base_extern_data_dim_refs` is never serialized.

What? This is exactly what this PR is about. Serializing the `_base_extern_data_dim_refs` in the specific network config by writing code that imports it from the base config.

**Comment by albertz** [comment]
> What? This is exactly what this PR is about. Serializing the `_base_extern_data_dim_refs` in the specific network config by writing code that imports it from the base config.

Ah sorry, I just read your description here and thought you just changed `set` to `list`.

**Comment by JackTemaki** [comment]
> > What? This is exactly what this PR is about. Serializing the `_base_extern_data_dim_refs` in the specific network config by writing code that imports it from the base config.
> 
> Ah sorry, I just read your description here and thought you just changed `set` to `list`.

Ah no, this was just meant as additional information.

File: nn/naming.py [hunk 1]
```diff
@@ -705,12 +710,19 @@ def get_ext_net_dict_py_code_str(self, root_module: nn.Module, *, with_imports:
...
     code_lines = []
+
     if with_imports:
       code_lines.append(self._ImportPyCodeStr + "\n")
     for import_str in imports:
       code_lines.append(import_str + "\n")
+
+    if ref_extern_data_dims_via_global_config:
+      code_lines += ["from returnn.config import get_global_config\n",
+                     "config = get_global_config()\n"]
```
Comment (albertz) [inline nn/naming.py:722]: Do not use hanging indents (same rule as in RETURNN).
```suggestion
      code_lines += [
        "from returnn.config import get_global_config\n",
        "config = get_global_config()\n"]
```

File: nn/naming.py [hunk 2]
```diff
@@ -705,12 +710,19 @@ def get_ext_net_dict_py_code_str(self, root_module: nn.Module, *, with_imports:
...
+      for value in self._base_extern_data_dim_refs:
+        code_lines += [f"{value.py_id_name()} = config.typed_dict[{value.py_id_name()!r}]\n"]
```
Comment (albertz) [inline nn/naming.py:724]:
```suggestion
        code_lines.append(f"{value.py_id_name()} = config.typed_dict[{value.py_id_name()!r}]\n")
```

**Review by albertz — APPROVED** [review]

**Pull request merged**